supervisor/src/bin/supervisor-encore.rs

```
@@ -0,0 +1,278 @@
+#[derive(serde::Serialize, serde::Deserialize)]
+struct RuntimeConfig {
+    #[serde(default)]
+    hosted_services: Vec<String>,
+    #[serde(default)]
+    gateways: Vec<GatewayConfig>,
+}
+
+#[derive(serde::Serialize, serde::Deserialize)]
+struct GatewayConfig {
+    name: String,
+}
+
+// Create a process config for a given set of services and gateways
+fn create_process_config(
+    services: Vec<String>,
+    gateways: Vec<String>,
+    port: u16,
+    service_ports: &HashMap<String, u16>,
+    cfg: &SupervisorConfig,
+) -> Result<Process, String> {
+    // Append all supervisor environment variables
+    let mut env = std::env::vars().collect::<HashMap<String, String>>();
+
+    // Find a process config that contains all the services and gateways
+    let binary_config = cfg
+        .procs
+        .iter()
+        .find(|p| {
+            (services.iter().all(|s| p.services.contains(s)))
+                && (gateways.iter().all(|g| p.gateways.contains(g)))
+        })
+        .ok_or_else(|| {
+            format!(
+                "No matching proc found for services {:?} gateways {:?}",
+                services, gateways
+            )
+        })?;
+
+    // Add proc-specific environment variables
+    env.extend(binary_config.env.iter().map(|e| {
+        let parts: Vec<&str> = e.splitn(2, '=').collect();
+        (
+            parts[0].to_string(),
+            parts.get(1).map_or("".to_string(), |&v| v.to_string()),
+        )
+    }));
+
+    // Add the port and process config to the environment
+    env.extend(vec![
+        ("PORT".to_string(), port.to_string()),
+        (
+            "ENCORE_PROCESS_CONFIG".to_string(),
+            base64::engine::general_purpose::STANDARD.encode(
+                serde_json::to_string(&ProcessConfig {
+                    hosted_gateways: gateways,
+                    hosted_services: services,
+                    local_service_ports: service_ports.clone(),
+                })
+                .unwrap(),
```

Not sure in which cases this can fail, maybe there are none? but if it can fail for some reason you could return an error here instead with `.map_err(|e| <format error somehow>)?` (can be made simpler by using `anyhow`)